### PR TITLE
Add support for sharing nodes across namespaces

### PR DESCRIPTION
**Pull request by juanfont**
This PR adds support for sharing nodes among namespaces - essentially implementing this https://tailscale.com/kb/1084/sharing/

Nodes from a given namespace can be 'shared' (thus becoming a peer) to another namespace. Routes are not shared, to replicate Tailscale.com behaviour.

### CODE REVIEW

File: sharing_nodes.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+type SharedNode struct {
+	gorm.Model
+	MachineID   uint64
+	Machine     Machine
+	NamespaceID uint
+	Namespace   Namespace
+}
+
+// ShareNodeInNamespace adds a machine as a shared node to a namespace
+func (h *Headscale) ShareNodeInNamespace(m *Machine, ns *Namespace) error {
```
Comment (kradalby) [inline sharing_nodes.go:18]: The wording of this is a bit confusing, it sounds like we are checking if it is in a namespace, e.g.
```go
if ShareNodeInNamespace() {}
```

How about:
`AddSharedNodeToNamespace`

Also we are getting quite inconsistent with `Machine` vs `Node`, Should this be `SharedMachine` ?

File: poll.go [hunk 2]
```diff
@@ -440,7 +440,7 @@ func (h *Headscale) scheduledPollWorker(
 		case <-updateCheckerTicker.C:
 			// Send an update request regardless of outdated or not, if data is sent
 			// to the node is determined in the updateChan consumer block
-			n, _ := m.toNode()
+			n, _ := m.toNode(true)
```
Comment (kradalby) [inline poll.go:443]: I really dont like that arguments cant have descriptors.

Comment (juanfont) [reply poll.go:443]: Me neither. Any idea here? I reckon we should keep Tailscale.com's behaviour of not exposing the routes across shared machines...

File: sharing_nodes.go (hunk 1 above)
Comment (juanfont) [reply sharing_nodes.go]: Changed to `SharedMachine`. It makes sense, having `Machine` as the main DB object.

File: cmd/headscale/cli/nodes.go [hunk 3]
```diff
@@ -79,17 +80,34 @@ var listNodesCmd = &cobra.Command{
 		if err != nil {
 			log.Fatalf("Error initializing: %s", err)
 		}
+
+		ns, err := h.GetNamespace(n)
```
Comment (kradalby) [inline cmd/headscale/cli/nodes.go:84]:
```suggestion
		namespace, err := h.GetNamespace(n)
```

File: cmd/headscale/cli/nodes.go [hunk 4]
```diff
@@ -79,17 +80,34 @@ var listNodesCmd = &cobra.Command{
...
+			JsonOutput(allMachines, err, o)
 			return
 		}
 
 		if err != nil {
 			log.Fatalf("Error getting nodes: %s", err)
 		}
 
-		d, err := nodesToPtables(*machines)
+		d, err := nodesToPtables(*ns, allMachines)
```
Comment (kradalby) [inline cmd/headscale/cli/nodes.go:110]:
```suggestion
		d, err := nodesToPtables(*namespace, allMachines)
```

File: cmd/headscale/cli/nodes.go [hunk 5]
```diff
@@ -145,8 +163,62 @@ var deleteNodeCmd = &cobra.Command{
 	},
 }
 
-func nodesToPtables(m []headscale.Machine) (pterm.TableData, error) {
-	d := pterm.TableData{{"ID", "Name", "NodeKey", "IP address", "Ephemeral", "Last seen", "Online"}}
+var shareNodeCmd = &cobra.Command{
```
Comment (kradalby) [inline cmd/headscale/cli/nodes.go:166]:
```suggestion
var shareMachineCmd = &cobra.Command{
```

File: cmd/headscale/cli/nodes.go [hunk 6]
```diff
@@ -145,8 +163,62 @@ var deleteNodeCmd = &cobra.Command{
...
+		id, err := strconv.Atoi(args[0])
+		if err != nil {
+			log.Fatalf("Error converting ID to integer: %s", err)
+		}
+		m, err := h.GetMachineByID(uint64(id))
+		if err != nil {
+			log.Fatalf("Error getting node: %s", err)
+		}
+
+		err = h.AddSharedMachineToNamespace(m, destNs)
```
Comment (kradalby) [inline cmd/headscale/cli/nodes.go:206]:
```suggestion
		err = h.AddSharedMachineToNamespace(machine, destinationNamespace)
```

File: cmd/headscale/cli/nodes.go [hunk 7]
```diff
@@ -145,8 +163,62 @@ var deleteNodeCmd = &cobra.Command{
...
+		if err != nil {
+			fmt.Printf("Error sharing node: %s\n", err)
+			return
+		}
+
+		fmt.Println("Node shared!")
+	},
+}
+
+func nodesToPtables(currNs headscale.Namespace, m []headscale.Machine) (pterm.TableData, error) {
```
Comment (kradalby) [inline cmd/headscale/cli/nodes.go:220]:
```suggestion
func nodesToPtables(currentNamespace headscale.Namespace, machines []headscale.Machine) (pterm.TableData, error) {
```

File: cmd/headscale/cli/nodes.go [hunk 8]
```diff
@@ -145,8 +163,62 @@ var deleteNodeCmd = &cobra.Command{
...
+	d := pterm.TableData{{"ID", "Name", "NodeKey", "Namespace", "IP address", "Ephemeral", "Last seen", "Online"}}
 
 	for _, m := range m {
```
Comment (kradalby) [inline cmd/headscale/cli/nodes.go:223]:
```suggestion
	for _, machine := range machines {
```

This seems unfortunate, overwriting the list in the closure.

File: cmd/headscale/cli/nodes.go [hunk 9]
```diff
@@ -169,7 +241,14 @@ func nodesToPtables(m []headscale.Machine) (pterm.TableData, error) {
...
 		}
-		d = append(d, []string{strconv.FormatUint(m.ID, 10), m.Name, nodeKey.ShortString(), m.IPAddress, strconv.FormatBool(ephemeral), lastSeen.Format("2006-01-02 15:04:05"), online})
+
+		var namespace string
+		if currNs.ID == m.NamespaceID {
+			namespace = pterm.LightMagenta(m.Namespace.Name)
+		} else {
+			namespace = pterm.LightYellow(m.Namespace.Name)
+		}
+		d = append(d, []string{strconv.FormatUint(m.ID, 10), m.Name, nodeKey.ShortString(), namespace, m.IPAddress, strconv.FormatBool(ephemeral), lastSeen.Format("2006-01-02 15:04:05"), online})
```
Comment (kradalby) [inline cmd/headscale/cli/nodes.go:251]:
```suggestion
		var namespace string
		if currentNamespace.ID == machine.NamespaceID {
			namespace = pterm.LightMagenta(machine.Namespace.Name)
		} else {
			namespace = pterm.LightYellow(machine.Namespace.Name)
		}
		d = append(d, []string{strconv.FormatUint(machine.ID, 10), machine.Name, nodeKey.ShortString(), namespace, machine.IPAddress, strconv.FormatBool(ephemeral), lastSeen.Format("2006-01-02 15:04:05"), online})
```

File: machine.go [hunk 10]
```diff
@@ -165,16 +174,31 @@ func (h *Headscale) getPeers(m Machine) (*[]*tailcfg.Node, error) {
...
+
 	machines := []Machine{}
 	if err := h.db.Where("namespace_id = ? AND machine_key <> ? AND registered",
 		m.NamespaceID, m.MachineKey).Find(&machines).Error; err != nil {
 		log.Error().Err(err).Msg("Error accessing db")
 		return nil, err
 	}
 
+	// We fetch here machines that are shared to the `Namespace` of the machine we are getting peers for
+	sharedNodes := []SharedMachine{}
```
Comment (kradalby) [inline machine.go:186]:
```suggestion
	sharedMachines := []SharedMachine{}
```

File: machine.go [hunk 11]
```diff
@@ -165,16 +174,31 @@ func (h *Headscale) getPeers(m Machine) (*[]*tailcfg.Node, error) {
...
+	if err := h.db.Preload("Namespace").Preload("Machine").Where("namespace_id = ?",
+		m.NamespaceID).Find(&sharedNodes).Error; err != nil {
```
Comment (kradalby) [inline machine.go:188]:
```suggestion
		m.NamespaceID).Find(&sharedMachines).Error; err != nil {
```

File: poll.go (hunk 2 above)
Comment (kradalby) [reply poll.go:443]: No, sadly, I think other projects use a list object and functions with name, but it seems too much of a hassle for this.

File: routes.go [hunk 12]
```diff
@@ -56,6 +56,7 @@ func (h *Headscale) GetEnabledNodeRoutes(namespace string, nodeName string) ([]n
 	return routes, nil
 }
 
+// IsNodeRouteEnabled checks if a certain route has been enabled
 func (h *Headscale) IsNodeRouteEnabled(namespace string, nodeName string, routeStr string) bool {
```
Comment (kradalby) [inline routes.go:60]: These functions should probably also be renames, but its probably not worth it.

I think the most important is that the database schema and models has the "correct" names so we dont have to do some major migration at some point.

File: sharing.go [hunk 13]
```diff
@@ -0,0 +1,37 @@
+package headscale
+
+import "gorm.io/gorm"
+
+const errorSameNamespace = Error("Destination namespace same as origin")
+const errorNodeAlreadyShared = Error("Node already shared to this namespace")
```
Comment (kradalby) [inline sharing.go:6]:
```suggestion
const errorMachineAlreadyShared = Error("Node already shared to this namespace")
```

File: sharing.go [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+	Namespace   Namespace
+}
+
+// AddSharedMachineToNamespace adds a machine as a shared node to a namespace
+func (h *Headscale) AddSharedMachineToNamespace(m *Machine, ns *Namespace) error {
+	if m.NamespaceID == ns.ID {
+		return errorSameNamespace
+	}
+
+	sn := SharedMachine{}
```
Comment (kradalby) [inline sharing.go:23]:
```suggestion
	sharedMachine := SharedMachine{}
```

File: sharing_test.go [hunk 15]
```diff
@@ -0,0 +1,359 @@
...
+	c.Assert(err, check.IsNil)
+
+	p1s, err := h.getPeers(m1)
+	c.Assert(err, check.IsNil)
+	c.Assert(len(*p1s), check.Equals, 0)
+
+	err = h.AddSharedMachineToNamespace(&m2, n1)
+	c.Assert(err, check.IsNil)
+	err = h.AddSharedMachineToNamespace(&m2, n1)
+	c.Assert(err, check.Equals, errorNodeAlreadyShared)
```
Comment (kradalby) [inline sharing_test.go:186]:
```suggestion
	c.Assert(err, check.Equals, errorMachineAlreadyShared)
```

File: sharing.go [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+		return errorNodeAlreadyShared
+	}
+
+	sn = SharedMachine{
+		MachineID:   m.ID,
+		Machine:     *m,
+		NamespaceID: ns.ID,
+		Namespace:   *ns,
+	}
+	h.db.Save(&sn)
```
Comment (kradalby) [inline sharing.go:34]:
```suggestion
	if err := h.db.Where("machine_id = ? AND namespace_id", m.ID, ns.ID).First(&sharedMachine).Error; err == nil {
		return errorMachineAlreadyShared
	}

	sharedMachine = SharedMachine{
		MachineID:   m.ID,
		Machine:     *m,
		NamespaceID: ns.ID,
		Namespace:   *ns,
	}
	h.db.Save(&sharedMachine)
```

**Review by kradalby — CHANGES_REQUESTED** [review]

**Comment by kradalby** [comment]
Overall very nice :) but using these one-two letter variable names makes it really hard to review.

**Comment by kradalby** [comment]
And as discussed, we should probably have the integration tests up and running for this. That can be done in this or a subsequent PR.

File: routes.go (hunk 12 above)
Comment (juanfont) [reply routes.go:60]: Fully agree. Machine and SharedMachine are now written in stone.

File: machine.go [hunk 17]
```diff
@@ -85,24 +85,26 @@ func (m Machine) toNode() (*tailcfg.Node, error) {
...
-	routesStr := []string{}
-	if len(m.EnabledRoutes) != 0 {
-		allwIps, err := m.EnabledRoutes.MarshalJSON()
-		if err != nil {
-			return nil, err
-		}
-		err = json.Unmarshal(allwIps, &routesStr)
-		if err != nil {
-			return nil, err
+	if includeRoutes {
```
[the remark this answers is not in the document]
Comment (juanfont) [reply machine.go:90]: Added comment on the function

**Review by juanfont — COMMENTED** [review]
Changes applied!

**Pull request merged**